test/vagrant-service-manager/proxy_test.rb

```diff
@@ -0,0 +1,109 @@
+require_relative '../test_helper'
+require 'vagrant/util/downloader'
+
+module VagrantPlugins
+  module ServiceManager
+    describe 'Proxy Test' do
+      before do
+        @machine = fake_machine
+        @ui = FakeUI.new
+        # set test path
+        @plugin_test_path = "#{@machine.env.data_dir}/service-manager/test"
+        ServiceManager.temp_dir = "#{@plugin_test_path}/temp"
+        ServiceManager.bin_dir = "#{@plugin_test_path}/bin"
+
+        @configs = { htpasswd_file: "#{ServiceManager.temp_dir}/.htpasswd" }
+      end
+
+      after do
+        FileUtils.rmtree(@plugin_test_path) if File.directory? @plugin_test_path
+      end
+
+      describe 'Unauthenticated proxy environment' do
+        before do
+          @string_stream = StringIO.new
+          @options = { type: :docker, '--cli-version' => '1.10.0' }
+          @handler = ADBDockerBinaryHandler.new(@machine, @machine.env, @options)
+        end
+
+        after do
+          Thread.kill @download_thread
+          Thread.kill @proxy_thread
+          @proxy_server.shutdown
+        end
+
+        it 'should send client binary download request' do
+          @handler.build_download_url
+          @handler.validate_url
+          @handler.build_archive_path
+          @handler.ensure_binary_and_temp_directories
+
+          @proxy_thread = Thread.new do
```

Comment:
Have the proxy server accept "expected urls" via a setter of some sorts. It keeps a list of expected URLs and when asked for it (via _assertExpectedRequestsReceived_) it checks this list against the log. In fact the log is in this case not even needed. You could check each URL as it arrives and mark expected URLs as received as you go. 

Also, you want to be able to do something like _expectUrlAndReturnOctetStream(url, path_to_file)_. There you say that if you receive a given URL, instead of letting it pass through, create your own HTTP response and return the content of the specified file as data. This way we can avoid network roundtrips for downloading binaries like _oc_.